### PR TITLE
feat: add OneRobotics A1 assets and Reach environments

### LINKED ISSUE #7095: [Proposal] Add OneRobotics A1 robotic arm asset to Isaac Lab
### Proposal

**OneRobotics** ([[www.onerobot.com](http://www.onerobot.com/)](http://www.onerobot.com/)) would like to contribute the **OneRobotics A1 robotic arm** to Isaac Lab as an officially supported or community-contributed robot asset.

The A1 is a 7-DoF fixed-base robotic arm. We currently have a working Isaac Lab integration available here:

https://github.com/katazen/onerobot_h1

The repository currently includes:

* the A1 robot description and mesh assets;
* an Isaac Lab `ArticulationCfg` for the A1 right arm;
* a manager-based end-effector pose reach task based on Isaac Lab's existing manipulation/reach framework;
* an RSL-RL training configuration;
* a MuJoCo sim-to-sim validation setup.

Our primary goal is to contribute the **A1 robot asset and its Isaac Lab configuration**. We are also happy to contribute the reach task as a reference environment if the maintainers consider it useful.

Before preparing the upstream PR, we would like to confirm the preferred contribution structure.

Specifically:

1. Should the OneRobotics A1 configuration be contributed under `isaaclab_contrib`, `isaaclab_assets`, or another location?
2. According to the current asset contribution guidelines, should we split the URDF/USD/mesh files into a dedicated asset repository before submitting the PR?
3. Would you prefer the initial contribution to contain only the robot asset/configuration, or should we also include the A1 reach environment as a reference task?

### Motivation

OneRobotics would like to make the A1 available to the Isaac Lab community so that researchers and developers can directly use the robot for manipulation, reinforcement learning, sim-to-sim, and sim-to-real research.

Providing an upstream Isaac Lab configuration would also give users a reproducible robot definition with validated articulation, actuator, and joint configuration instead of requiring each user to create their own integration.

The existing implementation has already been validated with an end-effector pose-reaching task in Isaac Lab, and we also maintain a MuJoCo sim-to-sim setup for cross-simulator validation.

Regarding the asset contribution process, we previously contacted NVIDIA about contributing the A1 asset and completed the agreement requested during that process. We are currently following up with our NVIDIA contact regarding the next asset-hosting/legal-approval step.

We are prepared to make the licensing scope for the A1 model assets explicit and compatible with Isaac Lab's asset contribution requirements before submitting the PR.

### Alternatives

An alternative is to continue maintaining the A1 integration purely as an external Isaac Lab project.

Another option would be to contribute only the A1 asset/configuration first and keep the reach task in our external repository.

We are happy to follow whichever structure is preferred by the Isaac Lab maintainers.

### Build Info

* Isaac Lab Version: target contribution against `release/3.0.0-beta2`
* Isaac Sim Version: target the Isaac Sim version supported by `release/3.0.0-beta2`

The current repository is an external Isaac Lab project and we will update it to the maintainer-recommended target before submitting the PR.

### Additional context

OneRobotics website:

http://www.onerobot.com/

Current A1 Isaac Lab integration:

https://github.com/katazen/onerobot_h1

The current implementation still contains some external-project/template structure that we intend to clean up before preparing the upstream PR, including package metadata, asset-path handling, and explicit separation of code and asset licensing.

We would prefer to confirm the desired upstream location and asset-hosting structure first so that we can avoid unnecessary restructuring.

### Checklist

* [x] I have checked that there is no existing issue specifically proposing the OneRobotics A1 robot asset.

### Acceptance Criteria

* [ ] Confirm the preferred upstream location for the OneRobotics A1 asset/configuration.
* [ ] Confirm the preferred asset-hosting and repository structure.
* [ ] Make the A1 asset licensing and redistribution terms explicit.
* [ ] Remove local filesystem dependencies and make the asset configuration portable.
* [ ] Add the A1 articulation configuration following Isaac Lab conventions.
* [ ] Add documentation and appropriate tests.
* [ ] Add the A1 reach environment if requested by the maintainers.

### PR DESCRIPTION
**Pull request by T1Amoo**
# Description

Adds the **OneRobotics A1** robotic arm assets/configurations and contributed Reach environments discussed in #7095.

This contribution includes two supported A1 configurations:

- the fixed-base **7-DoF OneRobotics A1 right arm**;
- the fixed-base **14-DoF OneRobotics A1 bimanual robot with stand**.

It also adds two Reach environments:

- `IsaacContrib-Reach-OneRobotics-A1`
- `IsaacContrib-Reach-OneRobotics-A1-Bimanual`

Both tasks reuse Isaac Lab's current Reach framework and provide RSL-RL PPO configurations.

The unimanual A1 remains a reusable standalone robot asset. In the Reach environment it is mounted directly on the standard Isaac Lab Reach table, similar to a conventional table-mounted manipulator, without introducing an additional pedestal or combined robot/table model.

The bimanual configuration uses the authoritative hardware-provided whole-robot URDF, including its stand and fixed shoulder mounts.

A small A1-specific MDP component generates kinematically reachable pose commands from the source joint limits, joint axes, and fixed transforms. The FK implementation is regression-tested against the spawned end-effector poses in simulation.

Related proposal: #7095

## Asset source and licensing

The review-stage robot assets are retrieved from the public OneRobotics asset repository:

https://github.com/katazen/onerobot_h1

An immutable review bundle is also available as:

[`onerobotics-a1-2026-review-v1`](https://github.com/katazen/onerobot_h1/releases/tag/onerobotics-a1-2026-review-v1)

The OneRobotics A1 URDF and mesh assets are licensed under **CC BY 4.0**.

The Isaac Lab integration code follows the repository's **BSD-3-Clause** contribution conventions.

NVIDIA legal approval and the final hosted asset URI are still pending. The current Git-backed asset loading path is intended for technical review and can be replaced with the final approved hosted asset path once that process is complete.

### Unimanual A1

- 7 actuated joints
- fixed-base configuration
- standard Isaac Lab Reach table
- 7-D joint-position action space
- `Link7` end effector
- FK-reachable pose targets
- 200 Hz simulation / 50 Hz policy control
- RSL-RL PPO configuration

The robot base placement is derived from the table geometry and A1 base geometry rather than visually approximated.

A deterministic sample of 4096 FK targets showed **98.56% of target end-effector positions above the tabletop**, so the existing reachable-target distribution was preserved without adding table-specific rejection sampling.

### Bimanual A1

- 14 actuated joints, ordered right arm then left arm
- fixed-base whole-robot model with stand
- independent right/left 7-D joint-position actions
- `Link_r7` and `Link_l7` end effectors
- independent FK-reachable targets for both arms
- simultaneous right-and-left success criterion
- 56-D policy observation
- RSL-RL PPO configuration

The bimanual FK preserves the source fixed shoulder transforms and arbitrary joint axes and is checked against the corresponding simulated end-effector poses.

## Validation

The final contribution was validated with:

- `./isaaclab.sh --format`
- `git diff --check`
- **7/7** focused A1 asset tests
- **2/2** unimanual Reach tests
- **3/3** bimanual Reach tests
- environment discovery for both task IDs
- public asset retrieval and source-fidelity checks
- zero-action and random-action runtime checks
- exact FK-to-sim end-effector regression checks

GPU finite-state validation:

- unimanual: **256 environments × 500 steps**
- bimanual: **256 environments × 500 steps**
- **128,000 transitions per task**
- observations, rewards, joint states, commands, and applied efforts remained finite

Short RSL-RL integration validation:

- unimanual: **256 environments × 5 iterations**
- bimanual: **256 environments × 5 iterations**
- **30,720 transitions per task**
- finite rollout / optimization metrics
- checkpoints generated successfully

These short runs validate the training integration path only; they are not intended as convergence or policy-quality claims.

No mass, inertia, collision geometry, visual geometry, kinematic transform, joint axis, or joint-limit changes were introduced to make training easier. Actuator limits and drive parameters are overlaid from confirmed OneRobotics hardware data because the CAD-exported URDF contains placeholder effort/velocity fields.

## Type of change

- New feature (non-breaking change which adds functionality)
- Documentation update

## Release backport

- [ ]  Backport this pull request to the active release branch after it merges into `develop`

### OneRobotics A1 unimanual Reach

截图 2026-08-31 14-54-17

### OneRobotics A1 bimanual Reach

截图 2026-08-31 14-51-58

## Checklist

Docker and GPU tests run on demand. Push the commits you want tested, then
comment `run-ci` on the pull request.

- [x] I have read and understood the [contribution guidelines](https://isaac-sim.github.io/IsaacLab/main/source/refs/contributing.html)
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my feature works
- [x] I have added a changelog fragment under `source/<pkg>/changelog.d/` for every touched package
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/isaaclab_assets/isaaclab_assets/robots/onerobotics.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+def _retrieve_a1_asset(relative_path: str) -> str:
+    """Retrieve an A1 asset, refreshing an older cache only when needed."""
+    try:
+        return retrieve_git_asset_path(_ONEROBOTICS_A1_ASSET_SOURCE, relative_path)
+    except FileNotFoundError:
+        return retrieve_git_asset_path(_ONEROBOTICS_A1_ASSET_SOURCE, relative_path, force_update=True)
+
+
+_ONEROBOTICS_A1_RIGHT_URDF_PATH = _retrieve_a1_asset("source/h1_reach/h1_reach/assets/urdf/A1_2026/a1_r.urdf")
+_ONEROBOTICS_A1_BIMANUAL_URDF_PATH = _retrieve_a1_asset(
```
[the remark this answers is not in the document]
Comment (uniquemelody) [reply source/isaaclab_assets/isaaclab_assets/robots/onerobotics.py]: Fixed in 73a01ca5c. Remote A1 retrieval now requires the immutable review-bundle commit 3db6f614c8443bd9396fcd9304cf64a519f98f2a, uses a revision-specific cache, and accepts only a full 40-character commit SHA. The local checkout override remains available for development.

File: source/isaaclab_assets/isaaclab_assets/robots/onerobotics.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+def _retrieve_a1_asset(relative_path: str) -> str:
+    """Retrieve an A1 asset, refreshing an older cache only when needed."""
+    try:
+        return retrieve_git_asset_path(_ONEROBOTICS_A1_ASSET_SOURCE, relative_path)
+    except FileNotFoundError:
+        return retrieve_git_asset_path(_ONEROBOTICS_A1_ASSET_SOURCE, relative_path, force_update=True)
+
+
+_ONEROBOTICS_A1_RIGHT_URDF_PATH = _retrieve_a1_asset("source/h1_reach/h1_reach/assets/urdf/A1_2026/a1_r.urdf")
```
[the remark this answers is not in the document]
Comment (uniquemelody) [reply source/isaaclab_assets/isaaclab_assets/robots/onerobotics.py]: Fixed in 73a01ca5c. A1 asset resolution is now deferred to the URDF spawn function, so importing the asset configuration does not run Git. First spawn resolves the exact pinned review-bundle commit. Regression coverage verifies both deferred resolution and pinned retrieval.

File: source/isaaclab_tasks/isaaclab_tasks/contrib/reach/config/onerobotics_a1/mdp/fk_pose_command.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            fixed_transform[:3, 3] = torch.tensor(xyz, dtype=torch.float64)
+        self._fixed_transform = fixed_transform.to(self.device)
+
+        # Keep the original single-arm computation as a separate exact path. In
+        # particular, do not replace its z-axis matrices with generic Rodrigues
+        # matrices: even equivalent arithmetic could perturb existing commands.
+        self._use_legacy_z_axis_fk = cfg.fixed_transform is None and all(len(entry) == 3 for entry in cfg.chain)
+
+    def _forward_kinematics(self, joint_pos: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
+        """Return the Link7 pose in the A1 base frame using XYZW quaternions."""
```
[the remark this answers is not in the document]
Comment (uniquemelody) [reply source/isaaclab_tasks/isaaclab_tasks/contrib/reach/config/onerobotics_a1/mdp/fk_pose_command.py]: Fixed in 73a01ca5c. The docstring now describes the configured end effector generically and documents Isaac Lab quaternion order as (w, x, y, z).

File: source/isaaclab_tasks/isaaclab_tasks/contrib/reach/config/onerobotics_a1/mdp/fk_pose_command_cfg.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+"""Left-arm chain from ``Link_l0`` through ``Link_l7`` in the bimanual source model."""
+
+
+@configclass
+class FkReachablePoseCommandCfg(UniformPoseCommandCfg):
+    """Configuration for pose targets sampled by A1 forward kinematics."""
+
+    class_type: type[FkReachablePoseCommand] | str = "{DIR}.fk_pose_command:FkReachablePoseCommand"
+
+    chain: list[KinematicChainEntry] = MISSING
```
[the remark this answers is not in the document]
Comment (uniquemelody) [reply source/isaaclab_tasks/isaaclab_tasks/contrib/reach/config/onerobotics_a1/mdp/fk_pose_command_cfg.py:142]: Fixed in 73a01ca5c. The public aliases and configuration fields now document tuple layout, parent/joint frames, translation [m], fixed-axis RPY [rad], and unitless joint-axis coordinates.

**Comment by maxkra15** [comment]
run-ci
